### ISSUE TITLE
Empty Image Markup Classes cause not uploading image from Froala editor

### ISSUE THREAD

**Issue by emzet** [octobercms/october#4579]
- OctoberCMS Build: 455
- PHP Version: 7.3.3
- Database Engine: 10.3.8-MariaDB-1:10.3.8+maria~xenial-log
- Plugins Installed: N/A

### Description:
If there are empty image markup classes, image upload from Froala WYSIWYG editor does not work, it throws JS error in console:

```
build-min.js?v455:formatted:13176 Uncaught TypeError: Cannot read property 'title' of null
    at FE.html (build-min.js?v455:formatted:13176)
    at _content (build-min.js?v455:formatted:8494)
    at _build (build-min.js?v455:formatted:8549)
    at Object.buildList (build-min.js?v455:formatted:8569)
    at U (build-min.js?v455:formatted:11894)
    at n (build-min.js?v455:formatted:11655)
    at HTMLImageElement.ne (build-min.js?v455:formatted:12505)
    at k (build-min.js?v455:formatted:11948)
    at HTMLImageElement.n (build-min.js?v455:formatted:12136)
    at HTMLImageElement.dispatch (jquery.min.js?v=455:2)
```
![js](https://user-images.githubusercontent.com/5151119/63880759-9adafe80-c9ce-11e9-976f-4c9b14918aee.PNG)

The problem is that this part of JS code is directly from [Froala](https://github.com/froala/wysiwyg-editor/blob/v2.9.3/js/plugins/image.min.js), so do not know if it is really OctoberCMS issue, but if I add some dummy class to list, there is not error in console and image uploads succesfully. 

Settings are stored in database in table `system_settings` under record `backend_editor_settings` as

```
"html_style_image": [
  {
    "class_label": "Foo",
    "class_name": "bar"
  }
]
```
which is then converted to JS object:

```
{
  bar: "Foo"
}
```

But if there are no classes it looks like:

```
"html_style_image": [
  []
]
```
which is then converted to JS object (first picture line 13172):

```
{
  "": null
}
```
Possible fix in OctoberCMS could be, that this strange object should be converted to empty JS object `{}`, which I tried in isolated function in console.

### Steps To Reproduce:
1. Delete all image markup classes (Settings -> Editor Settings -> Markup Classes)

![markup-classes](https://user-images.githubusercontent.com/5151119/63880720-85fe6b00-c9ce-11e9-9744-a97ad882e9c1.PNG)

2. Try to upload any image from Froala WYSIWYG editor (image will not be uploaded really, it is grayed out and error is thrown in console)

![froala](https://user-images.githubusercontent.com/5151119/63881001-2bb1da00-c9cf-11e9-86de-86fd94443341.PNG)

**Comment by w20k** [comment]
@emzet could you verify if it's still an issue in the latest build. Getting error for the title not being set, but it's purely `Froala` issue. And it doesn't stop you from inserting images.

**Comment by emzet** [comment]
@w20k I made further investigation, what I found is that it's not purely `Froala` issue. You can leave more empty rows in those tables (click on _Insert Row_ button):

![user-styles](https://user-images.githubusercontent.com/5151119/65599030-1e234c00-df9d-11e9-9aae-3e26d4c08197.PNG)

So those styles are then stored in database as empty arrays:

```
"html_style_image": [
  [],
  [],
  []
]
```

User defined styles are rendered as `data-image-styles` atrribute in partial template of Rich Editor here:

https://github.com/octobercms/october/blob/882ca6019c9c1b124e6e3f052b6aec02b6fa826b/modules/backend/formwidgets/richeditor/partials/_richeditor.htm#L19

Which  is then rendered by October as:

![rich-editor-ok](https://user-images.githubusercontent.com/5151119/65598255-4ad66400-df9b-11e9-99e7-a9897fa4b01a.PNG)

and if there are empty classes:

![rich-editor-ko](https://user-images.githubusercontent.com/5151119/65598262-4dd15480-df9b-11e9-8a52-d4da87a65268.PNG)

`$imageStyles` data for this atrribute are get from here:

https://github.com/octobercms/october/blob/882ca6019c9c1b124e6e3f052b6aec02b6fa826b/modules/backend/formwidgets/RichEditor.php#L97

and if we go deeper than from this method:

https://github.com/octobercms/october/blob/882ca6019c9c1b124e6e3f052b6aec02b6fa826b/modules/backend/models/EditorSetting.php#L122-L137

Variable `$value` gets value:

```
Array
(
    [0] => Array
        (
        )

    [1] => Array
        (
        )

    [2] => Array
        (
        )

)
```

which then steps in method `array_build` (passes the condition), which callback function returns `[NULL, NULL]` for each item, because `array_get` did not find `class_name` nor `class_label` on nested empty array. 

If we go further then on this [line](https://github.com/octobercms/library/blob/4bda08c826fc3ac1efba32fa8799d25ced14191f/src/Support/Arr.php#L27) is assigment:

```
$results[NULL] = NULL;
```

which always results into this:

```
array(1) {
  [""]=>  NULL
}
```

and this value is encoded to JSON `json_encode($imageStyles)` and rendered in partial template as data atrribute value as: 

```
{
  "": null
}
``` 

I think it could affect another user defined styles (_paragraph, link, table and table cell_), but it does not have to be a big issue there. 

Possible fix can be in `getConfiguredStyles` function, but the better solution would be if option of inserting new blank rows will be disabled directly on those UI data tables.

**Comment by w20k** [comment]
> Possible fix can be in getConfiguredStyles function, but the better solution would be if option of inserting new blank rows will be disabled directly on those UI data tables.

@emzet thanks for a tip, will look at it. Shouldn't be that difficult to add an option to ignore empty rows.

**Comment by emzet** [comment]
@w20k If there are not user defined classes - settings should be stored in database as:

```
"html_style_image": []
```

and then here add condition to check empty array:

https://github.com/octobercms/october/blob/882ca6019c9c1b124e6e3f052b6aec02b6fa826b/modules/backend/models/EditorSetting.php#L130

This would be the first part of fix. 

The second part (or question) is how to store those settings as only one empty array without those unnecessary nested empty arrays. I have not figured out this yet, but I think solution is to fix something in this [file](https://github.com/octobercms/october/blob/882ca6019c9c1b124e6e3f052b6aec02b6fa826b/modules/backend/widgets/table/assets/js/table.js) -- disable _Insert Row_ button if previous row is empty, and do not send data about empty rows in POST payload to backend...

**Comment by Samuell1** [comment]
You can prevent this by adding required on that inputs no?

**Comment by emzet** [comment]
@Samuell1 which inputs you mean? I think there are not any inputs in **DataTable** Widget, it is quite complex...

**Comment by Samuell1** [comment]
@emzet inputs in datatable. if they allow "required: true" that should do trick to prevent anyone to put empty row, but now iam not sure.

**Comment by bennothommo** [comment]
Fixed by #4725.
